### PR TITLE
516 connect donation to contracts

### LINKED ISSUE #516: Submit donation wallet interaction
**User Story:**
As a
I want to
So that

**Acceptance Criteria**
GIVEN
WHEN
THEN

**Product & Design Links:**

**Tech Details:**

**Open Questions:**

**Notes/Assumptions:**
- This includes writing test coverage for the functionality added in the ticket, if necessary

### PR DESCRIPTION
**Pull request by thelostone-mc**
##### Description

- added abi for round manager contracts (code ported from round manager)
- added useQFContext to handle progress modal states (with tests)
- updated ViewBallotPage to include the ProgressModal and ErrorModal

##### Refers/Fixes

closes #516

### CODE REVIEW

File: packages/grant-explorer/src/context/QFDonationContext.tsx [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+  token: PayoutToken,
+  amount: number,
+  context: QFDonationState
+): Promise<number> {
+  const { setTokenApprovalStatus } = context;
+
+  try {
+    setTokenApprovalStatus(ProgressStatus.IN_PROGRESS);
+
+    const votingStrategy = ethers.constants.AddressZero; // TODO wire in voting contract address
```
Comment (thelostone-mc) [inline packages/grant-explorer/src/context/QFDonationContext.tsx:178]: need to fetch from subgraph !
Not sure if we can reuse this from existing code

File: packages/grant-explorer/src/context/QFDonationContext.tsx [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+    // TODO: Skip if token is 0x0 as no approval is needed
```
Comment (thelostone-mc) [inline packages/grant-explorer/src/context/QFDonationContext.tsx:180]: Ideally if token is address 0x0 we don't need to do this step. 
This is a corner case which has to be handled

File: packages/grant-explorer/src/context/QFDonationContext.tsx [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+    // TODO : convert amount to BigNumber for approval
```
Comment (thelostone-mc) [inline packages/grant-explorer/src/context/QFDonationContext.tsx:182]: When approving -> my assumption is that this should be BigNumber and not a number. We likely have to do a converstion and multiply with token decimal

File: packages/grant-explorer/src/context/QFDonationContext.tsx [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+  }
+}
+
+function encodeQFVotes(donationToken: PayoutToken, donations: FinalBallotDonation[]) : BytesLike[] {
+  let encodedVotes: BytesLike[] = [];
+
+  donations.map(donation => {
+    const vote = [
+      donationToken.address,
+      donation.amount, // TODO: update to BigNum
```
Comment (thelostone-mc) [inline packages/grant-explorer/src/context/QFDonationContext.tsx:271]: We have to do a conversion and multiply with token decimal

File: packages/grant-explorer/src/context/QFDonationContext.tsx [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+      "0x0" // TODO: update to project address
```
Comment (thelostone-mc) [inline packages/grant-explorer/src/context/QFDonationContext.tsx:272]: Right now we store the project registry id but that isn't of much help. We need the payout address

File: packages/grant-explorer/src/context/QFDonationContext.tsx [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+  const context = useContext<QFDonationState>(
+    QFDonationContext
+  );
+  if (context === undefined) {
+    throw new Error(
+      "useQFDonation must be used within a QFDonationProvider"
+    );
+  }
+
+  const { signer } = useWallet(); // THIS BREAKS
```
Comment (thelostone-mc) [inline packages/grant-explorer/src/context/QFDonationContext.tsx:149]: This doesn't seem to work. 
I'm not sure what's causing this

File: packages/grant-explorer/src/features/ViewBallotPage.tsx [hunk 7]
```diff
@@ -62,6 +68,61 @@ export default function ViewBallot() {
...
+    ) {
+      setTimeout(() => {
+        setOpenProgressModal(false);
+        setOpenErrorModal(true);
+      }, modalDelayMs);
+    }
+
+    if (indexingStatus === ProgressStatus.IS_ERROR) {
+      setTimeout(() => {
+        navigate("/"); // TODO: Wire in thank you page
```
Comment (thelostone-mc) [inline packages/grant-explorer/src/features/ViewBallotPage.tsx:94]: This has to be wired in for next task

File: packages/grant-explorer/src/features/ViewBallotPage.tsx [hunk 8]
```diff
@@ -657,27 +718,37 @@ export default function ViewBallot() {
...
+        <ProgressModal
+          isOpen={openProgressModal}
+          subheading={"Please hold while we submit your donation."}
+          steps={progressSteps}
+        />
+        <ErrorModal
+          isOpen={openErrorModal}
+          setIsOpen={setOpenErrorModal}
+          // eslint-disable-next-line @typescript-eslint/no-empty-function
+          tryAgainFn={() => {}} // TODO: Wire in next story
```
Comment (thelostone-mc) [inline packages/grant-explorer/src/features/ViewBallotPage.tsx:749]: This has to be wired in for next task

**Review by bhargavaparoksham — APPROVED** [review]

**Review by nfrgosselin — APPROVED** [review]

**Review by melissa-neira — APPROVED** [review]

**Pull request merged**